gtn/graph.h

```diff
@@ -201,15 +200,18 @@ class Graph {
    * `Graph::numArcs()` elements.
    */
   float* weights() {
-    assert(sharedWeights_ != nullptr);
-    return sharedWeights_->data();
+    return const_cast<float*>(const_cast<const Graph*>(this)->weights());
```

Comment:
I don't think you need `const_cast<const Graph*>(this)` here, which is actually a no-op.
As far as I know, the purpose of `const_cast` is to remove `const`, not to add `const`.

Also, `this` can always be used to invoke a `const` member function.

Reply:
It needs a `static_cast` or a `const_cast` otherwise it will call the wrong version and recurse infinitely. This is a common idiom for implementing a non const member method in terms of a const member method. 

https://www.cppstories.com/2020/11/share-code-const-nonconst.html/

Reply:
Ah, I see. I was wrong. Thanks. I usually see people using `static_cast<const Graph*>(this)`